Require timestamps for importmulti keys

ryanofsky opened the pull request:
Require timestamps for importmulti keys

Additionally, accept a "now" timestamp, to allow avoiding rescans for keys
which are known never to have been used.

Note that the behavior when "now" is specified is slightly different than the
previous behavior when no timestamp was specified at all. Previously, when no
timestamp was specified, it would avoid rescanning during the importmulti call,
but set the key's nCreateTime value to 1, which would not prevent future block
reads in later ScanForWalletTransactions calls. With this change, passing a
"now" timestamp will set the key's nCreateTime to the current block time
instead of 1.

Fixes #9491

ryanofsky committed:
Require timestamps for importmulti keys

Additionally, accept a "now" timestamp, to allow avoiding rescans for keys
which are known never to have been used.

Note that the behavior when "now" is specified is slightly different than the
previous behavior when no timestamp was specified at all. Previously, when no
timestamp was specified, it would avoid rescanning during the importmulti call,
but set the key's nCreateTime value to 1, which would not prevent future block
reads in later ScanForWalletTransactions calls. With this change, passing a
"now" timestamp will set the key's nCreateTime to the current block time
instead of 1.

Fixes #9491

TheBlueMatt commented:
Needs 0.14 tag.

achow101 commented on src/wallet/rpcdump.cpp:
```diff
@@ -970,13 +984,16 @@ UniValue importmulti(const JSONRPCRequest& mainRequest)
             "  [     (array of json objects)\n"
             "    {\n"
             "      \"scriptPubKey\": \"<script>\" | { \"address\":\"<address>\" }, (string / json, required) Type of scriptPubKey (string for script, json for address)\n"
+            "      \"timestamp\": 1454686740 | \"now\"                       , (integer / string) Integer timestamp, or the string \"now\" to substitute the current synced\n"
```
You should say whether it is required or optional

jonasschnelli replied:
If we are at this, why do we have the number `1454686740` here? Can't we just say `<timestamp>`?

ryanofsky replied:
Updated in af06509f9c0bb92231ab41fd5fd9e89106b65b0e. Personally I think `"timestamp": 1454686740` is more informative than `"timestamp": <timestamp>`, made the requested change anyway.

gmaxwell commented:
Concept ACK but people with more opinions (and knowledge about json rpc norms) should comment on the type of the timestamp argument being variable.

jonasschnelli commented:
Concept ACK, will test.

I liked the idea from @sipa where he mentioned that "unknown" as timestamp for avoiding rescans seems to be the better choice (instead of "now"). "Now" would imply that the key was generated during import which is (almost) impossible.

MarcoFalke commented:
sipa proposed `timestamp: "unused"` instead of "now".
https://github.com/bitcoin/bitcoin/issues/9491#issuecomment-277054374

laanwj commented:
> Concept ACK but people with more opinions (and knowledge about json rpc norms) should comment on the type of the timestamp argument being variable.

It's hard to wrap in bindings for statically typed languages - could be worked around with an extra flag that specifies how to handle the timestamp.

ryanofsky committed:
Add test to check new importmulti "now" value

Easiest way to test this was to expose the timestamp via the validateaddress
RPC (which was already looking up and returning key metadata).

ryanofsky commented:
Added RPC test to verify that this is correctly interpreting the "now" value and saving the current time as the key creation time in 2cd875a53ed4ea7bde36f943603949f30bdcbc30.

ryanofsky commented:
Squashed 2cd875a53ed4ea7bde36f943603949f30bdcbc30 -> c07ebb6efced42cb9e4ed8bce387288ae705dca8 ([multinow.3](https://github.com/ryanofsky/bitcoin/commits/pr/multinow.3) -> [multinow.4](https://github.com/ryanofsky/bitcoin/commits/pr/multinow.4))

TheBlueMatt commented on src/wallet/rpcdump.cpp:
```diff
@@ -1015,6 +1032,12 @@ UniValue importmulti(const JSONRPCRequest& mainRequest)
     EnsureWalletIsUnlocked();
 
+    // Verify all timestamps are present before importing any keys.
+    const int64_t now = chainActive.Tip() ? chainActive.Tip()->GetBlockTime() : 0;
```
Can we use the MTP here?

ryanofsky replied:
Switched to MTP in dc402a4444d09263b74483a7d8ed904325dd89f6

TheBlueMatt commented:
utACK c07ebb6efced42cb9e4ed8bce387288ae705dca8 if we switch to using MTP for the "now" timestamp.

ryanofsky committed:
Use MTP for importmulti "now" timestamps

ryanofsky reviewed (commented):
Now using MTP in dc402a4444d09263b74483a7d8ed904325dd89f6

TheBlueMatt commented:
utACK dc402a4444d09263b74483a7d8ed904325dd89f6

laanwj commented:
> I liked the idea from @sipa where he mentioned that "unknown" as timestamp for avoiding rescans seems to be the better choice (instead of "now"). "Now" would imply that the key was generated during import which is (almost) impossible.
> sipa proposed `timestamp: "unused"` instead of "now".

A "unused" flag makes sense, however that is not a property of the timestamp.

kallewoof commented:
`Timestamp = unused` sounds strange to me as well. `Timestamp = now` makes more sense.

kallewoof commented on src/rpc/misc.cpp:
```diff
@@ -167,6 +167,7 @@ UniValue validateaddress(const JSONRPCRequest& request)
             "  \"pubkey\" : \"publickeyhex\",    (string) The hex value of the raw public key\n"
             "  \"iscompressed\" : true|false,  (boolean) If the address is compressed\n"
             "  \"account\" : \"account\"         (string) DEPRECATED. The account associated with the address, \"\" is the default account\n"
+            "  \"timestamp\" : <timestamp>,      (number, optional) The creation time of the key if available\n"
```
Nit: Non-string arguments tend to be represented as simply `name`, not `<name>`, e.g. https://github.com/bitcoin/bitcoin/blame/master/src/rpc/blockchain.cpp#L649

kallewoof commented on src/wallet/rpcdump.cpp:
```diff
@@ -640,7 +640,8 @@ UniValue dumpwallet(const JSONRPCRequest& request)
 
 
-UniValue processImport(const UniValue& data) {
+UniValue processImport(const UniValue& data, int64_t timestamp)
```
Nit: `const int64_t timestamp`
Nit 2 / coding standard: Functions in Bitcoin generally start with a capital letter, i.e. `ProcessImport` here. (Exception being the RPC commands, which match the CLI command names.) I know you didn't name this one, but it coincides with the new function below (`getImportTimestamp`). Renaming this one should not be a big issue, as it is used in one place only.

kallewoof commented on src/wallet/rpcdump.cpp:
```diff
@@ -958,6 +958,20 @@ UniValue processImport(const UniValue& data) {
     }
 }
 
+int64_t getImportTimestamp(const UniValue& data, int64_t now)
```
Nit / coding standard / repeat: `GetImportTimestamp` (see above)

kallewoof commented on src/wallet/rpcdump.cpp:
```diff
@@ -970,13 +984,16 @@ UniValue importmulti(const JSONRPCRequest& mainRequest)
             "  [     (array of json objects)\n"
             "    {\n"
             "      \"scriptPubKey\": \"<script>\" | { \"address\":\"<address>\" }, (string / json, required) Type of scriptPubKey (string for script, json for address)\n"
+            "      \"timestamp\": <timestamp> | \"now\"                      , (integer / string, required) Integer timestamp, or the string \"now\" to substitute the current synced\n"
```
Nit / repeat: `<timestamp>` -> `timestamp` (see above comment)

kallewoof commented on src/wallet/rpcdump.cpp:
```diff
@@ -1015,6 +1032,12 @@ UniValue importmulti(const JSONRPCRequest& mainRequest)
     EnsureWalletIsUnlocked();
 
+    // Verify all timestamps are present before importing any keys.
+    const int64_t now = chainActive.Tip() ? chainActive.Tip()->GetMedianTimePast() : 0;
```
I don't know if 0 is a good default for `now`, but it doesn't seem harmful either. I wonder if the function shouldn't simply throw a `runtime_exception` if `!chainActive.Tip()` saying to sync up first..

kallewoof replied:
OK, yeah I think a notification would be useful.

kallewoof commented on src/wallet/rpcdump.cpp:
```diff
@@ -970,13 +984,16 @@ UniValue importmulti(const JSONRPCRequest& mainRequest)
+            "      \"timestamp\": <timestamp> | \"now\"                      , (integer / string, required) Integer timestamp, or the string \"now\" to substitute the current synced\n"
+            "                                                              blockchain time. The timestamp of the oldest key will determine how far back blockchain rescans\n"
+            "                                                              for missing wallet transactions need to begin. \"now\" can be specified to bypass scanning, for\n"
+            "                                                              keys which are known to never have been used, and 0 can be specified to scan the entire blockchain.\n"
```
Language: "The timestamp of the oldest key will determine how far back blockchain rescans need to begin for finding missing wallet transactions." maybe?

kallewoof reviewed (approved):
utACK, did not review tests

ryanofsky commented:
Squashed 6524d401485a8dbd9881e6d69b3f542bb6598d6b -> 715f05d9987b6b9d8c2737d40a7113dd0680cfc4 ([multinow.6](https://github.com/ryanofsky/bitcoin/commits/pr/multinow.6) -> [multinow.7](https://github.com/ryanofsky/bitcoin/commits/pr/multinow.7))

jonasschnelli reviewed (approved):
Tested ACK 715f05d9987b6b9d8c2737d40a7113dd0680cfc4

ryanofsky commented:
Rebased 715f05d9987b6b9d8c2737d40a7113dd0680cfc4 -> 266a8114cbe2a87a6c84d7690a7716a18d782c56 ([multinow.7](https://github.com/ryanofsky/bitcoin/commits/pr/multinow.7) -> [multinow.8](https://github.com/ryanofsky/bitcoin/commits/pr/multinow.8)) following the #9227 merge.

This PR didn't conflict with #9227, but I'm about to rebase #9108 on top of this, and #9108 did conflict with #9227.

morcos commented:
utACK 266a811
